### ISSUE TITLE
Binary accuracy mismatch with train_on_batch accuracy

### ISSUE THREAD

**Issue by StatPal** [keras-team/keras#14761]
Please go to Stack Overflow for help and support:
https://stackoverflow.com/questions/tagged/keras

------------------------

### System information

-   **Have I written custom code (as opposed to using a stock example script provided in Keras)**: Yes
-   **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Manjaro Linux 21.0.5 (Ornara), and also Fedora release 34
-   **TensorFlow installed from (source or binary)**: binary (using conda)
-   **TensorFlow version (use command below)**: 2.4.1
-   **Python version**: 3.9.5
-   **GPU model and memory**: Quadro K2200 and NA
-   **Exact command to reproduce**: 

```python
import numpy as np
from keras.layers import Dense
from keras.models import Model, Sequential
from keras.optimizers import Adam
from keras.layers.advanced_activations import LeakyReLU

## Toy data generation: 
n = 80; p=2
X1 = np.concatenate([np.tile(np.array([10]), (n, p)), np.tile(np.array([20]), (n, p))])
y1 = np.concatenate([np.tile(np.array([0]), (n,)), np.tile(np.array([1]), (n,))])

## Seed:
seed_val = 12
import tensorflow as tf
import random as python_random
np.random.seed(seed_val)
python_random.seed(seed_val)
tf.random.set_seed(seed_val)


# define the keras model
def build_model(X, y):
    model = Sequential()
    model.add(Dense(12, input_dim=p, activation='relu'))
    model.add(LeakyReLU(alpha=0.01))
    model.add(Dense(1, activation='sigmoid'))
    print(model.summary())
    return model

## Build and compile
mod = build_model(X1, y1)
mod.compile(loss='binary_crossentropy', optimizer=Adam(learning_rate=0.005), metrics=[tf.keras.metrics.BinaryAccuracy()])


losses = []; accuracies = []; iteration_checkpoints = []

def train(X_train, y_train, iterations, batch_size, sample_interval):
    for iteration in range(iterations):
        idx = np.random.randint(0, X_train.shape[0], batch_size)
        X_batch = X_train[idx]
        y_batch = y_train[idx]
        
        ## Accuracy before
        m = tf.keras.metrics.BinaryAccuracy()
        m.update_state(np.expand_dims(y_batch, axis=0), mod.predict(X_batch))
        
        ## Train on batch:
        d_loss, accuracy = mod.train_on_batch(X_batch, y_batch)
        
        ## Accuracy after
        m2 = tf.keras.metrics.BinaryAccuracy()
        m2.update_state(np.expand_dims(y_batch, axis=0), mod.predict(X_batch))
        
        print("%d> Before train: %.5f,\t After train: %.5f,\t\t acc: %.5f%%" %(iteration+1, m.result().numpy(), m2.result().numpy(), accuracy*100))
        
        if (iteration + 1) % sample_interval == 0:
            # Save losses and accuracies so they can be plotted after training
            losses.append([d_loss, accuracy]); accuracies.append(100.0 * accuracy); iteration_checkpoints.append(iteration + 1)


model_hist = train(X1, y1, iterations=20, batch_size=32, sample_interval=1)
```

### Describe the problem
**Binary accuracy in a toy problem is not giving the same accuracy as the accuracy provided by `train_on_batch`.** 
This toy example captures shows that the BinaryAccuracy before and after the batch training. It does not always match with any of them. A reproducible example is provided.  

The output has a line:
```
17> Before train: 0.53125,	 After train: 0.62500,		 acc: 100.00000%
```
Before train is the `result()` of `tf.keras.metrics.BinaryAccuracy` before training in each step. 
After train is the `result()` of `tf.keras.metrics.BinaryAccuracy` after training in each step. 
acc is the accuracy generated by `train_on_batch` command. 
(Increasing the `iterations` in the last line of the code gives more such phenomenon.)

### Source code / logs
This section contains the whole output of the program and versions of the packages managed by conda. 

#### Output of the program contains:
```
2021-06-19 12:40:03.144136: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-06-19 12:40:03.144320: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  SSE4.1 SSE4.2 AVX AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-06-19 12:40:03.145882: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 2. Tune using inter_op_parallelism_threads for best performance.
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
dense (Dense)                (None, 12)                36        
_________________________________________________________________
leaky_re_lu (LeakyReLU)      (None, 12)                0         
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 13        
=================================================================
Total params: 49
Trainable params: 49
Non-trainable params: 0
_________________________________________________________________
None
2021-06-19 12:40:03.204089: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2021-06-19 12:40:03.219466: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 3093880000 Hz
1> Before train: 0.65625,	 After train: 0.65625,		 acc: 65.62500%
2> Before train: 0.46875,	 After train: 0.46875,		 acc: 46.87500%
3> Before train: 0.43750,	 After train: 0.43750,		 acc: 43.75000%
4> Before train: 0.50000,	 After train: 0.50000,		 acc: 50.00000%
5> Before train: 0.46875,	 After train: 0.46875,		 acc: 46.87500%
6> Before train: 0.46875,	 After train: 0.46875,		 acc: 46.87500%
7> Before train: 0.37500,	 After train: 0.37500,		 acc: 37.50000%
8> Before train: 0.50000,	 After train: 0.50000,		 acc: 50.00000%
9> Before train: 0.53125,	 After train: 0.53125,		 acc: 53.12500%
10> Before train: 0.56250,	 After train: 0.56250,		 acc: 56.25000%
11> Before train: 0.65625,	 After train: 0.65625,		 acc: 65.62500%
12> Before train: 0.46875,	 After train: 0.46875,		 acc: 46.87500%
13> Before train: 0.43750,	 After train: 0.43750,		 acc: 43.75000%
14> Before train: 0.59375,	 After train: 0.59375,		 acc: 59.37500%
15> Before train: 0.53125,	 After train: 0.53125,		 acc: 53.12500%
16> Before train: 0.50000,	 After train: 0.50000,		 acc: 50.00000%
17> Before train: 0.53125,	 After train: 0.62500,		 acc: 100.00000%
18> Before train: 0.50000,	 After train: 0.50000,		 acc: 50.00000%
19> Before train: 0.53125,	 After train: 0.53125,		 acc: 53.12500%
20> Before train: 0.56250,	 After train: 0.56250,		 acc: 56.25000%
```

#### Versions managed by conda:
The existing versions managed by conda 
(generated using `conda list --export`,
and virtual en can be generated using the channel `conda-forge`
`conda create --name workingenv --file <following as a file>  -c conda-forge`)
```
# This file may be used to create an environment using:
# $ conda create --name <env> --file <this file>
# platform: linux-64
_libgcc_mutex=0.1=main
_openmp_mutex=4.5=1_gnu
_tflow_select=2.3.0=mkl
absl-py=0.12.0=py39h06a4308_0
aiohttp=3.7.4=py39h27cfd23_1
astor=0.8.1=py39h06a4308_0
astunparse=1.6.3=py_0
async-timeout=3.0.1=py39h06a4308_0
attrs=21.2.0=pyhd3eb1b0_0
blas=1.0=mkl
blinker=1.4=py39h06a4308_0
brotlipy=0.7.0=py39h27cfd23_1003
c-ares=1.17.1=h27cfd23_0
ca-certificates=2021.5.25=h06a4308_1
cachetools=4.2.2=pyhd3eb1b0_0
certifi=2021.5.30=py39h06a4308_0
cffi=1.14.5=py39h261ae71_0
chardet=3.0.4=py39h06a4308_1003
click=8.0.1=pyhd3eb1b0_0
coverage=5.5=py39h27cfd23_2
cryptography=3.4.7=py39hd23ed53_0
cycler=0.10.0=py39h06a4308_0
cython=0.29.23=py39h2531618_0
dbus=1.13.18=hb2f20db_0
expat=2.4.1=h2531618_2
fontconfig=2.13.1=h6c09931_0
freetype=2.10.4=h5ab3b9f_0
gast=0.4.0=py_0
glib=2.68.2=h36276a3_0
google-auth=1.31.0=pyhd3eb1b0_0
google-auth-oauthlib=0.4.1=py_2
google-pasta=0.2.0=py_0
grpcio=1.36.1=py39h2157cd5_1
gst-plugins-base=1.14.0=h8213a91_2
gstreamer=1.14.0=h28cd5cc_2
h5py=2.10.0=py39hec9cf62_0
hdf5=1.10.6=hb1b8bf9_0
icu=58.2=he6710b0_3
idna=2.10=pyhd3eb1b0_0
importlib-metadata=3.10.0=py39h06a4308_0
intel-openmp=2021.2.0=h06a4308_610
jpeg=9b=h024ee3a_2
keras=2.4.3=0
keras-base=2.4.3=py_0
keras-preprocessing=1.1.2=pyhd3eb1b0_0
kiwisolver=1.3.1=py39h2531618_0
lcms2=2.12=h3be6417_0
ld_impl_linux-64=2.35.1=h7274673_9
libffi=3.3=he6710b0_2
libgcc-ng=9.3.0=h5101ec6_17
libgfortran-ng=7.5.0=ha8ba4b0_17
libgfortran4=7.5.0=ha8ba4b0_17
libgomp=9.3.0=h5101ec6_17
libpng=1.6.37=hbc83047_0
libprotobuf=3.14.0=h8c45485_0
libstdcxx-ng=9.3.0=hd4cf53a_17
libtiff=4.2.0=h85742a9_0
libuuid=1.0.3=h1bed415_2
libwebp-base=1.2.0=h27cfd23_0
libxcb=1.14=h7b6447c_0
libxml2=2.9.10=hb55368b_3
lz4-c=1.9.3=h2531618_0
markdown=3.3.4=py39h06a4308_0
matplotlib=3.3.4=py39h06a4308_0
matplotlib-base=3.3.4=py39h62a2d02_0
mkl=2021.2.0=h06a4308_296
mkl-service=2.3.0=py39h27cfd23_1
mkl_fft=1.3.0=py39h42c9631_2
mkl_random=1.2.1=py39ha9443f7_2
multidict=5.1.0=py39h27cfd23_2
ncurses=6.2=he6710b0_1
nibabel=3.2.1=pyhd8ed1ab_0
numpy=1.20.2=py39h2d18471_0
numpy-base=1.20.2=py39hfae3a4d_0
oauthlib=3.1.1=pyhd3eb1b0_0
olefile=0.46=py_0
openssl=1.1.1k=h27cfd23_0
opt_einsum=3.3.0=pyhd3eb1b0_1
packaging=20.9=pyhd3eb1b0_0
pcre=8.44=he6710b0_0
pillow=8.2.0=py39he98fc37_0
pip=21.1.2=py39h06a4308_0
protobuf=3.14.0=py39h2531618_1
pyasn1=0.4.8=py_0
pyasn1-modules=0.2.8=py_0
pycparser=2.20=py_2
pydicom=2.1.2=pyhd3deb0d_0
pydot=1.2.4=py_0
pyjwt=2.1.0=py39h06a4308_0
pyopenssl=20.0.1=pyhd3eb1b0_1
pyparsing=2.4.7=pyhd3eb1b0_0
pyqt=5.9.2=py39h2531618_6
pysocks=1.7.1=py39h06a4308_0
python=3.9.5=h12debd9_4
python-dateutil=2.8.1=pyhd3eb1b0_0
python-flatbuffers=1.12=pyhd3eb1b0_0
pyyaml=5.4.1=py39h27cfd23_1
qt=5.9.7=h5867ecd_1
readline=8.1=h27cfd23_0
requests=2.25.1=pyhd3eb1b0_0
requests-oauthlib=1.3.0=py_0
rsa=4.7.2=pyhd3eb1b0_1
scipy=1.6.2=py39had2a1c9_1
setuptools=52.0.0=py39h06a4308_0
sip=4.19.13=py39h2531618_0
six=1.16.0=pyhd3eb1b0_0
sqlite=3.35.4=hdfb4753_0
tensorboard=2.5.0=py_0
tensorboard-plugin-wit=1.6.0=py_0
tensorflow=2.4.1=mkl_py39h4683426_0
tensorflow-base=2.4.1=mkl_py39h43e0292_0
tensorflow-estimator=2.5.0=pyh7b7c402_0
termcolor=1.1.0=py39h06a4308_1
tk=8.6.10=hbc83047_0
tornado=6.1=py39h27cfd23_0
typing-extensions=3.7.4.3=hd3eb1b0_0
typing_extensions=3.7.4.3=pyh06a4308_0
tzdata=2020f=h52ac0ba_0
urllib3=1.26.4=pyhd3eb1b0_0
werkzeug=1.0.1=pyhd3eb1b0_0
wheel=0.36.2=pyhd3eb1b0_0
wrapt=1.12.1=py39he8ac12f_1
xz=5.2.5=h7b6447c_0
yaml=0.2.5=h7b6447c_0
yarl=1.6.3=py39h27cfd23_0
zipp=3.4.1=pyhd3eb1b0_0
zlib=1.2.11=h7b6447c_3
zstd=1.4.9=haebb681_0
```

**Comment by jvishnuvardhan** [comment]
I can reproduce the issue with `tf-nightly`. [Here](https://colab.research.google.com/gist/jvishnuvardhan/f91a728b9447ea1b7bbfb09f2e28d8a0/untitled1070.ipynb) is a gist for our reference. Thanks!

**Comment by LukeWood** [comment]
will dig into this!

**Comment by jvishnuvardhan** [comment]
@LukeWood @StatPal I think I found the root-cause of the behavior you are noticing.

[train_on_batch](https://www.tensorflow.org/api_docs/python/tf/keras/Model?version=nightly#train_on_batch) has an input `reset_metrics`=True (by default). Which means the metric is reset for every batch. Hence, the accuracy value is for a batch.

reset_metrics | If True, the metrics returned will be only for this batch. If False, the metrics will be statefully accumulated across batches.
-- | --

However, [`model.update_state()`](https://www.tensorflow.org/api_docs/python/tf/keras/metrics/Metric?version=nightly#update_state), accumulate the accuracy over all batches.

So, in order to get similar results, you need to add `reset_metrics`= False to the `train_on_batch` as shown below.

`d_loss, accuracy = mod.train_on_batch(X_batch, y_batch,reset_metrics=False)`

Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/0dd58b70db91e0ed54f2ea74f780c4c4/untitled1070.ipynb)

**Comment by jvishnuvardhan** [comment]
I think this is intended behavior. Only one batch had 100% accuracy (batch metric) as each of the elements `y_pred` and `y_true` are matching for that batch only. 

After making the change I listed above, cumulative accuracy from both the approaches are very close. Thanks!

@StatPal I will close the issue as this was resolved. Feel free to reopen the issue if this persists again. Thanks!
